File: flare/cutoffs.py
"""
The cutoffs module gives a few different options for smoothly sending the GP
kernel to zero near the boundary of the cutoff sphere.
"""
from math import cos, sin, pi
from numba import njit


@njit
def hard_cutoff(r_cut: float, ri: float, ci: float):
    """A hard cutoff that assigns a value of 1 to all interatomic distances.

    Args:
        r_cut (float): Cutoff value (in angstrom).
        ri (float): Interatomic distance.
        ci (float): Cartesian coordinate divided by the distance.

    Returns:
        (float, float): Cutoff value and its derivative.
    """
    return 1, 0

@njit
def quadratic_cutoff_bound(r_cut: float, ri: float, ci: float):
    """A quadratic cutoff that goes to zero smoothly at the cutoff boundary.

    Args:
        r_cut (float): Cutoff value (in angstrom).
        ri (float): Interatomic distance.
        ci (float): Cartesian coordinate divided by the distance.

    Returns:
        (float, float): Cutoff value and its derivative.
    """

    if (r_cut > ri):
        rdiff = r_cut - ri
        fi = rdiff * rdiff
        fdi = 2 * rdiff * ci
    else:
        fi = 0
        fdi = 0

    return fi, fdi


@njit
def quadratic_cutoff(r_cut: float, ri: float, ci: float):
    """A quadratic cutoff that goes to zero smoothly at the cutoff boundary.

    Args:
        r_cut (float): Cutoff value (in angstrom).
        ri (float): Interatomic distance.
        ci (float): Cartesian coordinate divided by the distance.

    Returns:
        (float, float): Cutoff value and its derivative.
    """

    rdiff = r_cut - ri
    fi = rdiff * rdiff
    fdi = 2 * rdiff * ci

    return fi, fdi


@njit
def cubic_cutoff(r_cut: float, ri: float, ci: float):
    """A cubic cutoff that goes to zero smoothly at the cutoff boundary.

    Args:
        r_cut (float): Cutoff value (in angstrom).
        ri (float): Interatomic distance.
        ci (float): Cartesian coordinate divided by the distance.

    Returns:
        (float, float): Cutoff value and its derivative.
    """

    rdiff = r_cut - ri
    fi = rdiff * rdiff * rdiff
    fdi = 3 * rdiff * rdiff * ci

    return fi, fdi


@njit
def cosine_cutoff(r_cut: float, ri: float, ci: float, d: float = 1):
    """A cosine cutoff that returns 1 up to r_cut - d, and assigns a cosine
    envelope to values of r between r_cut - d and r_cut. Based on Eq. 24 of
    Albert P. Bartók and Gábor Csányi. "Gaussian approximation potentials: A
    brief tutorial introduction." International Journal of Quantum Chemistry
    115.16 (2015): 1051-1057.

    Args:
        r_cut (float): Cutoff value (in angstrom).
        ri (float): Interatomic distance.
        ci (float): Cartesian coordinate divided by the distance.

    Returns:
        (float, float): Cutoff value and its derivative.
    """

    if ri > r_cut - d:
        fi = (1/2) * (cos(pi * (ri - r_cut + d) / d) + 1)
        fdi = (pi/(2 * d)) * sin(pi * (r_cut - ri) / d) * ci
    else:
        fi = 1
        fdi = 0

    return fi, fdi

File: src/cutoffs.cpp
#include "ace.h"
#include <cmath>
#define Pi 3.14159265358979323846

void quadratic_cutoff(double * rcut_vals, double r, double rcut,
                      std::vector<double> cutoff_hyps){
                
    if (r > rcut){
        rcut_vals[0] = 0;
        rcut_vals[1] = 0;
        return;
    }

    double rdiff = r - rcut;
    rcut_vals[0] = rdiff * rdiff;
    rcut_vals[1] = 2 * rdiff;
}

void cos_cutoff(double * rcut_vals, double r, double rcut,
                std::vector<double> cutoff_hyps){

    // Calculate the cosine cutoff function and its gradient. If r > rcut, the array is returned unchanged.
    if (r > rcut){
        rcut_vals[0] = 0;
        rcut_vals[1] = 0;
        return;
    }

    double cutoff_val = (1./2.) * (cos(Pi * r / rcut) + 1);
    double cutoff_derv = -Pi * sin(Pi * r / rcut) / (2 * rcut);

    rcut_vals[0] = cutoff_val;
    rcut_vals[1] = cutoff_derv;
}

void hard_cutoff(double * rcut_vals, double r, double rcut,
                 std::vector<double> cutoff_hyps){
    if (r > rcut){
        rcut_vals[0] = 0;
        rcut_vals[1] = 0;
        return;
    }

    rcut_vals[0] = 1;
    rcut_vals[1] = 0;

}
